python/eos/analysis.py

```diff
@@ -425,10 +435,15 @@ def sample_pmc(self, log_proposal, step_N=1000, steps=10, final_N=5000, rng=np.r
         samples = np.apply_along_axis(self._x_to_par, 1, sampler.samples[:])  # Rescale the parameters back to their original bounds
         weights = sampler.weights[:][:, 0]
         adjusted_weights = np.copy(weights)
-        for i, w in enumerate(adjusted_weights):
-            if w <= 0 or np.isnan(w):
-                adjusted_weights[i] = eps
-        normalized_weights = adjusted_weights / np.sum(adjusted_weights)
+        normalized_weights = adjusted_weights / np.sum(
+            adjusted_weights,
+            where = np.logical_or(adjusted_weights > 0, np.isfinite(adjusted_weights))
+        ) # sum positive finite weights only
+        normalized_weights = np.where(
+                np.logical_or(normalized_weights <= 0, np.isnan(normalized_weights)),
+                eps,
+                normalized_weights
+        ) # replace negative and nan weights by eps
```

Comment:
Seeing the same code thrice, should we make it a separaste function? Also, is there any benefit to using [pypmc's function ```pypmc.tools.perp```](https://pypmc.github.io/api.html#pypmc.tools.convergence.perp)?

A quick look at that source code seems better for the zero weights, but has not a workable solution for NaN and negative weights.

Reply:
Indeed the pypmc's `perp` doesn't treat nan weights, that are even propagated due to the normalization. We could write our own `perp` function.